File: platform/Flutter/sweeteditor/lib/widget/sweet_editor_controller.dart
part of '../sweeteditor.dart';

class SweetEditorController {
  _SweetEditorWidgetState? _state;
  final EditorEventBus _eventBus = EditorEventBus();
  final List<VoidCallback> _readyCallbacks = <VoidCallback>[];
  bool _associationEstablished = false;
  bool _terminated = false;

  void _attach(_SweetEditorWidgetState state) {
    if (_associationEstablished) {
      if (!identical(_state, state)) {
        throw StateError(
          'SweetEditorController cannot be rebound to another editor instance',
        );
      }
      return;
    }
    _associationEstablished = true;
    _state = state;
    final callbacks = List<VoidCallback>.from(_readyCallbacks);
    _readyCallbacks.clear();
    for (final callback in callbacks) {
      callback();
    }
  }

  void _detach() {
    _state = null;
    _terminated = true;
    _readyCallbacks.clear();
  }

  bool get isAttached => _state != null;

  core.EditorCore? get _editorCore => _state?._editorCore;

  void _withEditorCore(void Function(core.EditorCore editorCore) action) {
    if (_terminated) return;
    final editorCore = _editorCore;
    if (editorCore == null) return;
    action(editorCore);
  }

  void whenReady(VoidCallback callback) {
    if (_state != null) {
      callback();
      return;
    }
    if (_terminated) return;
    _readyCallbacks.add(callback);
  }

  void loadDocument(core.Document document) {
    if (_terminated) return;
    _state?._loadDocument(document);
  }

  void loadText(String text) {
    if (_terminated) return;
    _state?._loadText(text);
  }

  core.Document? getDocument() => _state?._document;

  String getContent() => _state?._getContent() ?? '';
  int get lineCount => _state?._document?.lineCount ?? 0;
  String getLineText(int line) => _state?._document?.getLineText(line) ?? '';

  EditorSettings? get settings => getSettings();

  EditorSettings? getSettings() => _state?._session.settings;

  LanguageConfiguration? get languageConfiguration =>
      getLanguageConfiguration();

  set languageConfiguration(LanguageConfiguration? value) =>
      setLanguageConfiguration(value);

  LanguageConfiguration? getLanguageConfiguration() =>
      _state?._session.languageConfiguration;

  void setLanguageConfiguration(LanguageConfiguration? value) {
    if (_terminated) return;
    _state?._applyLanguageConfiguration(value);
  }

  EditorMetadata? get metadata => getMetadata();

  set metadata(EditorMetadata? value) => setMetadata(value);

  EditorMetadata? getMetadata() => _state?._session.metadata;

  void setMetadata(EditorMetadata? value) {
    if (_terminated) return;
    _state?._applyMetadata(value);
  }

  core.TextPosition getCursorPosition() =>
      _state?._editorCore?.getCursorPosition() ?? const core.TextPosition(0, 0);

  void setCursorPosition(Object positionOrLine, [int? column]) {
    final position = _resolveTextPositionArgument(
      positionOrLine,
      column,
      methodName: 'setCursorPosition',
    );
    _state?._editorCore?.setCursorPosition(position.line, position.column);
    _state?._flush();
  }

  void gotoPosition(int line, int column) {
    _state?._editorCore?.gotoPosition(line, column);
    _state?._flush();
  }

  core.TextRange? getSelection() => _state?._editorCore?.getSelection();

  void setSelection(
    int startLine,
    int startColumn,
    int endLine,
    int endColumn,
  ) {
    _state?._editorCore?.setSelection(
      startLine,
      startColumn,
      endLine,
      endColumn,
    );
    _state?._flush();
  }

  void selectAll() {
    _state?._interactionController.selectAll();
  }

  String getSelectedText() => _state?._editorCore?.getSelectedText() ?? '';

  void insertText(String text) {
    _state?._interactionController.insertText(text);
  }

  void replaceText(
    Object rangeOrStartLine,
    Object textOrStartColumn, [
    int? endLine,
    int? endColumn,
    String? text,
  ]) {
    final args = _resolveReplaceTextArguments(
      rangeOrStartLine,
      textOrStartColumn,
      endLine,
      endColumn,
      text,
    );
    _state?._interactionController.replaceText(args.$1, args.$2);
  }

  void deleteText(
    Object rangeOrStartLine, [
    int? startColumn,
    int? endLine,
    int? endColumn,
  ]) {
    final range = _resolveTextRangeArgument(
      rangeOrStartLine,
      startColumn,
      endLine,
      endColumn,
      methodName: 'deleteText',
    );
    _state?._interactionController.deleteText(range);
  }

  void insertSnippet(String snippetTemplate) {
    _state?._interactionController.insertSnippet(snippetTemplate);
  }

  void moveLineUp() {
    _withEditorCore((editorCore) {
      final result = editorCore.moveLineUp();
      _state?._interactionController.dispatchTextChangedForController(
        TextChangeAction.insert,
        result,
      );
    });
  }

  void moveLineDown() {
    _withEditorCore((editorCore) {
      final result = editorCore.moveLineDown();
      _state?._interactionController.dispatchTextChangedForController(
        TextChangeAction.insert,
        result,
      );
    });
  }

  void copyLineUp() {
    _withEditorCore((editorCore) {
      final result = editorCore.copyLineUp();
      _state?._interactionController.dispatchTextChangedForController(
        TextChangeAction.insert,
        result,
      );
    });
  }

  void copyLineDown() {
    _withEditorCore((editorCore) {
      final result = editorCore.copyLineDown();
      _state?._interactionController.dispatchTextChangedForController(
        TextChangeAction.insert,
        result,
      );
    });
  }

  void deleteLine() {
    _withEditorCore((editorCore) {
      final result = editorCore.deleteLine();
      _state?._interactionController.dispatchTextChangedForController(
        TextChangeAction.delete_,
        result,
      );
    });
  }

  void insertLineAbove() {
    _withEditorCore((editorCore) {
      final result = editorCore.insertLineAbove();
      _state?._interactionController.dispatchTextChangedForController(
        TextChangeAction.insert,
        result,
      );
    });
  }

  void insertLineBelow() {
    _withEditorCore((editorCore) {
      final result = editorCore.insertLineBelow();
      _state?._interactionController.dispatchTextChangedForController(
        TextChangeAction.insert,
        result,
      );
    });
  }

  void undo() {
    _state?._interactionController.undo();
  }

  void redo() {
    _state?._interactionController.redo();
  }

  bool get canUndo => _state?._editorCore?.canUndo ?? false;
  bool get canRedo => _state?._editorCore?.canRedo ?? false;

  core.TextRange getWordRangeAtCursor() =>
      _state?._editorCore?.getWordRangeAtCursor() ??
      const core.TextRange(core.TextPosition(0, 0), core.TextPosition(0, 0));

  String getWordAtCursor() => _state?._editorCore?.getWordAtCursor() ?? '';

  void addCompletionProvider(CompletionProvider provider) =>
      _state?._completionProviderManager.addProvider(provider);

  void removeCompletionProvider(CompletionProvider provider) =>
      _state?._completionProviderManager.removeProvider(provider);

  void addDecorationProvider(DecorationProvider provider) =>
      _state?._decorationProviderManager.addProvider(provider);

  void removeDecorationProvider(DecorationProvider provider) =>
      _state?._decorationProviderManager.removeProvider(provider);

  void requestDecorationRefresh() =>
      _state?._decorationProviderManager.requestRefresh();

  void addNewLineActionProvider(NewLineActionProvider provider) =>
      _state?._newLineActionProviderManager.addProvider(provider);

  void removeNewLineActionProvider(NewLineActionProvider provider) =>
      _state?._newLineActionProviderManager.removeProvider(provider);

  void triggerCompletion() => _state?._completionProviderManager
      .triggerCompletion(CompletionTriggerKind.invoked, null);

  void showCompletionItems(List<CompletionItem> items) =>
      _state?._completionProviderManager.showItems(items);

  void dismissCompletion() => _state?._completionProviderManager.dismiss();

  void setCompletionItemRenderer(CompletionItemViewBuilder? renderer) =>
      _state?._completionPopupController.setViewBuilder(renderer);

  bool get isCompletionShowing =>
      _state?._completionPopupController.isShowing ?? false;

  void showInlineSuggestion(InlineSuggestion suggestion) =>
      _state?._inlineSuggestionController.show(suggestion);

  void dismissInlineSuggestion() =>
      _state?._inlineSuggestionController.dismiss();

  bool get isInlineSuggestionShowing =>
      _state?._inlineSuggestionController.isShowing ?? false;

  void setInlineSuggestionListener(InlineSuggestionListener? listener) =>
      _state?._inlineSuggestionController.setListener(listener);

  bool get hasSelection => _state?._editorCore?.getSelection() != null;

  void setSelectionMenuItemProvider(SelectionMenuItemProvider? provider) =>
      _state?._selectionMenuController.setItemProvider(provider);

  Stream<TextChangedEvent> get onTextChanged =>
      _eventBus.on<TextChangedEvent>();

  Stream<CursorChangedEvent> get onCursorChanged =>
      _eventBus.on<CursorChangedEvent>();

  Stream<SelectionChangedEvent> get onSelectionChanged =>
      _eventBus.on<SelectionChangedEvent>();

  Stream<ScrollChangedEvent> get onScrollChanged =>
      _eventBus.on<ScrollChangedEvent>();

  Stream<ScaleChangedEvent> get onScaleChanged =>
      _eventBus.on<ScaleChangedEvent>();

  Stream<LongPressEvent> get onLongPress => _eventBus.on<LongPressEvent>();

  Stream<DoubleTapEvent> get onDoubleTap => _eventBus.on<DoubleTapEvent>();

  Stream<ContextMenuEvent> get onContextMenu =>
      _eventBus.on<ContextMenuEvent>();

  Stream<GutterIconClickEvent> get onGutterIconClick =>
      _eventBus.on<GutterIconClickEvent>();

  Stream<InlayHintClickEvent> get onInlayHintClick =>
      _eventBus.on<InlayHintClickEvent>();

  Stream<CodeLensClickEvent> get onCodeLensClick =>
      _eventBus.on<CodeLensClickEvent>();

  Stream<LinkClickEvent> get onLinkClick => _eventBus.on<LinkClickEvent>();

  Stream<FoldToggleEvent> get onFoldToggle => _eventBus.on<FoldToggleEvent>();

  Stream<DocumentLoadedEvent> get onDocumentLoaded =>
      _eventBus.on<DocumentLoadedEvent>();

  Stream<SelectionMenuItemClickEvent> get onSelectionMenuItemClick =>
      _eventBus.on<SelectionMenuItemClickEvent>();

  void toggleFoldAt(int line) {
    _state?._editorCore?.toggleFoldAt(line);
    _state?._flush();
  }

  void foldAt(int line) {
    _state?._editorCore?.foldAt(line);
    _state?._flush();
  }

  void unfoldAt(int line) {
    _state?._editorCore?.unfoldAt(line);
    _state?._flush();
  }

  void foldAll() {
    _state?._editorCore?.foldAll();
    _state?._flush();
  }

  void unfoldAll() {
    _state?._editorCore?.unfoldAll();
    _state?._flush();
  }

  core.ScrollMetrics getScrollMetrics() =>
      _state?._editorCore?.getScrollMetrics() ?? core.ScrollMetrics.empty;

  void setScroll(double scrollX, double scrollY) {
    _state?._editorCore?.setScroll(scrollX, scrollY);
    _state?._flush();
  }

  core.CursorRect getPositionRect(int line, int column) =>
      _state?._editorCore?.getPositionRect(line, column) ??
      const core.CursorRect();

  core.CursorRect getCursorRect() =>
      _state?._editorCore?.getCursorRect() ?? const core.CursorRect();

  core.IntRange getVisibleLineRange() {
    final editorCore = _state?._editorCore;
    if (editorCore == null) {
      return const core.IntRange(0, -1);
    }
    if (_state?._session.renderModel.visualLines.isEmpty ?? true) {
      editorCore.buildRenderModel();
    }
    return editorCore.getVisibleLineRange();
  }

  int getTotalLineCount() => _state?._document?.lineCount ?? 0;

  void scrollToLine(
    int line, {
    core.ScrollBehavior behavior = core.ScrollBehavior.center,
  }) {
    _state?._editorCore?.scrollToLine(line, behavior: behavior);
    _state?._flush();
  }

  bool isLineVisible(int line) =>
      _state?._editorCore?.isLineVisible(line) ?? true;

  int get totalLineCount => getTotalLineCount();

  EditorKeyMap getKeyMap() =>
      _state?._session.keyMap ?? EditorKeyMap.defaultKeyMap();

  void setKeyMap(EditorKeyMap keyMap) {
    if (_terminated) return;
    _state?._applyKeyMap(keyMap);
  }

  void setEditorIconProvider(EditorIconProvider? provider) {
    if (_terminated) return;
    _state?._applyIconProvider(provider);
  }

  void applyTheme(EditorTheme theme) {
    if (_terminated) return;
    _state?._applyTheme(theme);
  }

  void setTheme(EditorTheme theme) => applyTheme(theme);

  EditorTheme? getTheme() => _state?._theme;

  void registerTextStyle(
    int styleId,
    int color, {
    int backgroundColor = 0,
    int fontStyle = 0,
  }) {
    _withEditorCore(
      (editorCore) => editorCore.registerTextStyle(
        styleId,
        color,
        backgroundColor: backgroundColor,
        fontStyle: fontStyle,
      ),
    );
  }

  void registerBatchTextStyles(Map<int, core.TextStyle> stylesById) {
    _withEditorCore((editorCore) {
      editorCore.registerBatchTextStyles(stylesById);
    });
  }

  void setLineSpans(
    int line,
    core.SpanLayer layer,
    List<core.StyleSpan> spans,
  ) {
    _withEditorCore((editorCore) {
      editorCore.setLineSpans(line, layer, spans);
    });
  }

  void setBatchLineSpans(
    core.SpanLayer layer,
    Map<int, List<core.StyleSpan>> spansByLine,
  ) {
    _withEditorCore((editorCore) {
      editorCore.setBatchLineSpans(layer, spansByLine);
    });
  }

  void setLineInlayHints(int line, List<core.InlayHint> hints) {
    _withEditorCore((editorCore) {
      editorCore.setLineInlayHints(line, hints);
    });
  }

  void setBatchLineInlayHints(Map<int, List<core.InlayHint>> hintsByLine) {
    _withEditorCore((editorCore) {
      editorCore.setBatchLineInlayHints(hintsByLine);
    });
  }

  void setLinePhantomTexts(int line, List<core.PhantomText> phantoms) {
    _withEditorCore((editorCore) {
      editorCore.setLinePhantomTexts(line, phantoms);
    });
  }

  void setBatchLinePhantomTexts(
    Map<int, List<core.PhantomText>> phantomsByLine,
  ) {
    _withEditorCore((editorCore) {
      editorCore.setBatchLinePhantomTexts(phantomsByLine);
    });
  }

  void setLineGutterIcons(int line, List<core.GutterIcon> icons) {
    _withEditorCore((editorCore) {
      editorCore.setLineGutterIcons(line, icons);
    });
  }

  void setBatchLineGutterIcons(Map<int, List<core.GutterIcon>> iconsByLine) {
    _withEditorCore((editorCore) {
      editorCore.setBatchLineGutterIcons(iconsByLine);
    });
  }

  void setLineCodeLens(int line, List<core.CodeLensItem> items) {
    _withEditorCore((editorCore) {
      editorCore.setLineCodeLens(line, items);
    });
  }

  void setBatchLineCodeLens(Map<int, List<core.CodeLensItem>> itemsByLine) {
    _withEditorCore((editorCore) {
      editorCore.setBatchLineCodeLens(itemsByLine);
    });
  }

  void setLineLinks(int line, List<core.LinkSpan> links) {
    _withEditorCore((editorCore) {
      editorCore.setLineLinks(line, links);
    });
  }

  void setBatchLineLinks(Map<int, List<core.LinkSpan>> linksByLine) {
    _withEditorCore((editorCore) {
      editorCore.setBatchLineLinks(linksByLine);
    });
  }

  String getLinkTargetAt(int line, int column) =>
      _state?._editorCore?.getLinkTargetAt(line, column) ?? '';

  void setLineDiagnostics(int line, List<core.Diagnostic> items) {
    _withEditorCore((editorCore) {
      editorCore.setLineDiagnostics(line, items);
    });
  }

  void setBatchLineDiagnostics(Map<int, List<core.Diagnostic>> itemsByLine) {
    _withEditorCore((editorCore) {
      editorCore.setBatchLineDiagnostics(itemsByLine);
    });
  }

  void setIndentGuides(List<core.IndentGuide> guides) {
    _withEditorCore((editorCore) {
      editorCore.setIndentGuides(guides);
    });
  }

  void setBracketGuides(List<core.BracketGuide> guides) {
    _withEditorCore((editorCore) {
      editorCore.setBracketGuides(guides);
    });
  }

  void setFlowGuides(List<core.FlowGuide> guides) {
    _withEditorCore((editorCore) {
      editorCore.setFlowGuides(guides);
    });
  }

  void setSeparatorGuides(List<core.SeparatorGuide> guides) {
    _withEditorCore((editorCore) {
      editorCore.setSeparatorGuides(guides);
    });
  }

  void setFoldRegions(List<core.FoldRegion> regions) {
    _withEditorCore((editorCore) {
      editorCore.setFoldRegions(regions);
    });
  }

  void setMatchedBrackets(
    int openLine,
    int openColumn,
    int closeLine,
    int closeColumn,
  ) {
    _withEditorCore((editorCore) {
      editorCore.setMatchedBrackets(
        openLine,
        openColumn,
        closeLine,
        closeColumn,
      );
    });
  }

  void clearMatchedBrackets() {
    _withEditorCore((editorCore) => editorCore.clearMatchedBrackets());
  }

  void clearHighlights([core.SpanLayer? layer]) {
    _withEditorCore((editorCore) {
      editorCore.clearHighlights(layer);
    });
  }

  void clearInlayHints() {
    _withEditorCore((editorCore) => editorCore.clearInlayHints());
  }

  void clearPhantomTexts() {
    _withEditorCore((editorCore) => editorCore.clearPhantomTexts());
  }

  void clearGutterIcons() {
    _withEditorCore((editorCore) => editorCore.clearGutterIcons());
  }

  void clearCodeLens() {
    _withEditorCore((editorCore) => editorCore.clearCodeLens());
  }

  void clearLinks() {
    _withEditorCore((editorCore) => editorCore.clearLinks());
  }

  void clearGuides() {
    _withEditorCore((editorCore) => editorCore.clearGuides());
  }

  void clearDiagnostics() {
    _withEditorCore((editorCore) => editorCore.clearDiagnostics());
  }

  void clearAllDecorations() {
    _withEditorCore((editorCore) => editorCore.clearAllDecorations());
  }

  void flush() => _state?._flush();

  core.TextPosition _resolveTextPositionArgument(
    Object positionOrLine,
    int? column, {
    required String methodName,
  }) {
    if (positionOrLine is core.TextPosition && column == null) {
      return positionOrLine;
    }
    if (positionOrLine is int && column != null) {
      return core.TextPosition(positionOrLine, column);
    }
    throw ArgumentError(
      '$methodName expects either a TextPosition or line and column integers.',
    );
  }

  core.TextRange _resolveTextRangeArgument(
    Object rangeOrStartLine,
    int? startColumn,
    int? endLine,
    int? endColumn, {
    required String methodName,
  }) {
    if (rangeOrStartLine is core.TextRange &&
        startColumn == null &&
        endLine == null &&
        endColumn == null) {
      return rangeOrStartLine;
    }
    if (rangeOrStartLine is int &&
        startColumn != null &&
        endLine != null &&
        endColumn != null) {
      return core.TextRange(
        core.TextPosition(rangeOrStartLine, startColumn),
        core.TextPosition(endLine, endColumn),
      );
    }
    throw ArgumentError(
      '$methodName expects either a TextRange or four integer coordinates.',
    );
  }

  (core.TextRange, String) _resolveReplaceTextArguments(
    Object rangeOrStartLine,
    Object textOrStartColumn,
    int? endLine,
    int? endColumn,
    String? text,
  ) {
    if (rangeOrStartLine is core.TextRange &&
        textOrStartColumn is String &&
        endLine == null &&
        endColumn == null &&
        text == null) {
      return (rangeOrStartLine, textOrStartColumn);
    }
    if (rangeOrStartLine is int &&
        textOrStartColumn is int &&
        endLine != null &&
        endColumn != null &&
        text != null) {
      return (
        core.TextRange(
          core.TextPosition(rangeOrStartLine, textOrStartColumn),
          core.TextPosition(endLine, endColumn),
        ),
        text,
      );
    }
    throw ArgumentError(
      'replaceText expects either (TextRange, String) or '
      '(int, int, int, int, String).',
    );
  }
}

File: platform/Emscripten/sdk/packages/widget/src/sweet-editor-controller.ts
import type { IAnyValue, ITextPosition, ITextRange } from "@sweeteditor/core";
import type { SweetEditorWidget } from "./legacy/sweet-editor-widget-legacy.js";

type IWidgetOp = (widget: SweetEditorWidget) => void;

export class SweetEditorController {
  private _widget: SweetEditorWidget | null = null;
  private _disposed = false;
  private _queuedOps: IWidgetOp[] = [];
  private _readyCallbacks: Array<(widget: SweetEditorWidget) => void> = [];

  whenReady(callback:(widget: SweetEditorWidget) => void) {
    if (typeof callback !== "function" || this._disposed) {
      return;
    }
    if (this._widget) {
      callback(this._widget);
      return;
    }
    this._readyCallbacks.push(callback);
  }

  bind(editorApi:SweetEditorWidget) {
    if (this._disposed || !editorApi) {
      return;
    }
    if (this._widget && this._widget !== editorApi) {
      throw new Error("SweetEditorController is already bound to another editor instance.");
    }
    this._widget = editorApi;

    const queued = this._queuedOps;
    this._queuedOps = [];
    queued.forEach((op) => op(editorApi));

    const callbacks = this._readyCallbacks;
    this._readyCallbacks = [];
    callbacks.forEach((callback) => callback(editorApi));
  }

  unbind() {
    this._widget = null;
  }

  dispose() {
    if (this._disposed) {
      return;
    }
    this._disposed = true;
    this._widget = null;
    this._queuedOps = [];
    this._readyCallbacks = [];
  }

  getNativeWidget() {
    return this._widget;
  }

  invoke(method:string, ...args:IAnyValue[]) {
    this._enqueueOrRun((widget) => {
      const fn = (widget as IAnyValue)?.[method];
      if (typeof fn === "function") {
        fn.apply(widget, args);
      }
    });
  }

  loadDocument(document:IAnyValue) { this.invoke("loadDocument", document); }
  getDocument() { return this._read("getDocument", null); }
  loadText(text:string, options:IAnyValue = {}) { this.invoke("loadText", text, options); }
  getText() { return this._read("getText", ""); }

  applyTheme(theme:IAnyValue = {}) { this.invoke("applyTheme", theme); }
  getTheme() { return this._read("getTheme", {}); }
  getSettings() { return this._read("getSettings", null); }
  setKeyMap(keyMap:IAnyValue) { this.invoke("setKeyMap", keyMap); }
  setTabSize(tabSize:number) { this.invoke("setTabSize", tabSize); }
  setBackspaceUnindent(enabled:boolean) { this.invoke("setBackspaceUnindent", enabled); }
  setInsertSpaces(enabled:boolean) { this.invoke("setInsertSpaces", enabled); }
  setCompositionEnabled(enabled:boolean) { this.invoke("setCompositionEnabled", enabled); }
  isCompositionEnabled() { return this._read("isCompositionEnabled", true); }
  openImeCandidate() { this.invoke("openImeCandidate"); }
  closeImeCandidate() { this.invoke("closeImeCandidate"); }
  getKeyMap() { return this._read("getKeyMap", null); }
  setEditorIconProvider(provider:IAnyValue) { this.invoke("setEditorIconProvider", provider); }
  getEditorIconProvider() { return this._read("getEditorIconProvider", null); }

  insertText(text:string) { this.invoke("insertText", text); }
  replaceText(range:ITextRange, text:string) { this.invoke("replaceText", range, text); }
  deleteText(range:ITextRange) { this.invoke("deleteText", range); }
  moveLineUp() { this.invoke("moveLineUp"); }
  moveLineDown() { this.invoke("moveLineDown"); }
  copyLineUp() { this.invoke("copyLineUp"); }
  copyLineDown() { this.invoke("copyLineDown"); }
  deleteLine() { this.invoke("deleteLine"); }
  insertLineAbove() { this.invoke("insertLineAbove"); }
  insertLineBelow() { this.invoke("insertLineBelow"); }
  undo() { this.invoke("undo"); }
  redo() { this.invoke("redo"); }
  canUndo() { return this._read("canUndo", false); }
  canRedo() { return this._read("canRedo", false); }
  copyToClipboard() { this.invoke("copyToClipboard"); }
  pasteFromClipboard() { this.invoke("pasteFromClipboard"); }
  cutToClipboard() { this.invoke("cutToClipboard"); }

  selectAll() { this.invoke("selectAll"); }
  getSelectedText() { return this._read("getSelectedText", ""); }
  setSelection(startOrRange:ITextRange | ITextPosition, startColumn?: number, endLine?: number, endColumn?: number) {
    this.invoke("setSelection", startOrRange, startColumn, endLine, endColumn);
  }
  getSelection() { return this._read("getSelection", null); }
  setCursorPosition(position:ITextPosition) { this.invoke("setCursorPosition", position); }
  getCursorPosition() { return this._read("getCursorPosition", null); }
  getWordRangeAtCursor() { return this._read("getWordRangeAtCursor", null); }
  getWordAtCursor() { return this._read("getWordAtCursor", ""); }
  moveCursorPageUp(extendSelection:boolean = false) { this.invoke("moveCursorPageUp", extendSelection); }
  moveCursorPageDown(extendSelection:boolean = false) { this.invoke("moveCursorPageDown", extendSelection); }

  gotoPosition(line:number, column:number) { this.invoke("gotoPosition", line, column); }
  scrollToLine(line:number, behavior:number = 0) { this.invoke("scrollToLine", line, behavior); }
  setScroll(x:number, y:number) { this.invoke("setScroll", x, y); }
  ensureCursorVisible() { this.invoke("ensureCursorVisible"); }
  stopFling() { this.invoke("stopFling"); }
  getScrollMetrics() { return this._read("getScrollMetrics", null); }
  getPositionRect(line:number, column:number) { return this._read("getPositionRect", null, line, column); }
  getCursorRect() { return this._read("getCursorRect", null); }

  toggleFoldAt(line:number) { this.invoke("toggleFoldAt", line); }
  foldAt(line:number) { this.invoke("foldAt", line); }
  unfoldAt(line:number) { this.invoke("unfoldAt", line); }
  foldAll() { this.invoke("foldAll"); }
  unfoldAll() { this.invoke("unfoldAll"); }
  isLineVisible(line:number) { return this._read("isLineVisible", false, line); }

  setLanguageConfiguration(config:IAnyValue) { this.invoke("setLanguageConfiguration", config); }
  setAutoClosingPairs(pairs:IAnyValue) { this.invoke("setAutoClosingPairs", pairs); }
  getLanguageConfiguration() { return this._read("getLanguageConfiguration", null); }
  setMetadata(metadata:IAnyValue) { this.invoke("setMetadata", metadata); }
  getMetadata() { return this._read("getMetadata", null); }

  addDecorationProvider(provider:IAnyValue) { this.invoke("addDecorationProvider", provider); }
  removeDecorationProvider(provider:IAnyValue) { this.invoke("removeDecorationProvider", provider); }
  requestDecorationRefresh() { this.invoke("requestDecorationRefresh"); }
  addCompletionProvider(provider:IAnyValue) { this.invoke("addCompletionProvider", provider); }
  removeCompletionProvider(provider:IAnyValue) { this.invoke("removeCompletionProvider", provider); }
  addNewLineActionProvider(provider:IAnyValue) { this.invoke("addNewLineActionProvider", provider); }
  removeNewLineActionProvider(provider:IAnyValue) { this.invoke("removeNewLineActionProvider", provider); }
  triggerCompletion() { this.invoke("triggerCompletion"); }
  showCompletionItems(items:IAnyValue[]) { this.invoke("showCompletionItems", items); }
  dismissCompletion() { this.invoke("dismissCompletion"); }
  setCompletionItemRenderer(renderer:IAnyValue) { this.invoke("setCompletionItemRenderer", renderer); }

  registerTextStyle(styleId:number, color:number, backgroundColor:number = 0, fontStyle:number = 0) {
    this.invoke("registerTextStyle", styleId, color, backgroundColor, fontStyle);
  }
  registerBatchTextStyles(stylesById:IAnyValue) { this.invoke("registerBatchTextStyles", stylesById); }
  setLineSpans(line:number, layer:IAnyValue, spans:IAnyValue[]) { this.invoke("setLineSpans", line, layer, spans); }
  setBatchLineSpans(layer:IAnyValue, spansByLine:IAnyValue) { this.invoke("setBatchLineSpans", layer, spansByLine); }
  setLineInlayHints(line:number, hints:IAnyValue[]) { this.invoke("setLineInlayHints", line, hints); }
  setBatchLineInlayHints(hintsByLine:IAnyValue) { this.invoke("setBatchLineInlayHints", hintsByLine); }
  setLinePhantomTexts(line:number, phantoms:IAnyValue) { this.invoke("setLinePhantomTexts", line, phantoms); }
  setBatchLinePhantomTexts(phantomsByLine:IAnyValue) { this.invoke("setBatchLinePhantomTexts", phantomsByLine); }
  setLineGutterIcons(line:number, icons:IAnyValue[]) { this.invoke("setLineGutterIcons", line, icons); }
  setBatchLineGutterIcons(iconsByLine:IAnyValue) { this.invoke("setBatchLineGutterIcons", iconsByLine); }
  setLineCodeLens(line:number, items:IAnyValue[]) { this.invoke("setLineCodeLens", line, items); }
  setBatchLineCodeLens(itemsByLine:IAnyValue) { this.invoke("setBatchLineCodeLens", itemsByLine); }
  clearCodeLens() { this.invoke("clearCodeLens"); }
  setLineLinks(line:number, links:IAnyValue[]) { this.invoke("setLineLinks", line, links); }
  setBatchLineLinks(linksByLine:IAnyValue) { this.invoke("setBatchLineLinks", linksByLine); }
  clearLinks() { this.invoke("clearLinks"); }
  getLinkTargetAt(line:number, column:number) { return this._read("getLinkTargetAt", "", line, column); }
  setLineDiagnostics(line:number, diagnostics:IAnyValue[]) { this.invoke("setLineDiagnostics", line, diagnostics); }
  setBatchLineDiagnostics(diagsByLine:IAnyValue) { this.invoke("setBatchLineDiagnostics", diagsByLine); }
  setIndentGuides(guides:IAnyValue[]) { this.invoke("setIndentGuides", guides); }
  setBracketGuides(guides:IAnyValue[]) { this.invoke("setBracketGuides", guides); }
  setFlowGuides(guides:IAnyValue[]) { this.invoke("setFlowGuides", guides); }
  setSeparatorGuides(guides:IAnyValue[]) { this.invoke("setSeparatorGuides", guides); }
  setFoldRegions(regions:IAnyValue[]) { this.invoke("setFoldRegions", regions); }
  clearHighlights(layer?:IAnyValue) { this.invoke("clearHighlights", layer); }
  clearInlayHints() { this.invoke("clearInlayHints"); }
  clearPhantomTexts() { this.invoke("clearPhantomTexts"); }
  clearGutterIcons() { this.invoke("clearGutterIcons"); }
  clearDiagnostics() { this.invoke("clearDiagnostics"); }
  clearGuides() { this.invoke("clearGuides"); }
  clearAllDecorations() { this.invoke("clearAllDecorations"); }
  flush() { this.invoke("flush"); }
  getVisibleLineRange(options:IAnyValue = {}) { return this._read("getVisibleLineRange", { start: 0, end: -1 }, options); }
  getTotalLineCount() { return this._read("getTotalLineCount", 0); }

  private _enqueueOrRun(action:IWidgetOp) {
    if (this._disposed) {
      return;
    }
    if (this._widget) {
      action(this._widget);
      return;
    }
    this._queuedOps.push(action);
  }

  private _read<T>(method:string, fallback:T, ...args:IAnyValue[]): T {
    const widget = this._widget;
    if (!widget || this._disposed) {
      return fallback;
    }
    const fn = (widget as IAnyValue)?.[method];
    if (typeof fn !== "function") {
      return fallback;
    }
    return fn.apply(widget, args) as T;
  }
}
